backend/src/apiserver/resource/resource_manager.go

```diff
@@ -195,616 +510,1147 @@ func (r *ResourceManager) ArchiveExperiment(ctx context.Context, experimentId st
 	if err != nil {
-		return util.Wrap(err, "Retry run failed")
+		return util.Wrapf(err, "Failed to retry run %s due to error fetching the run", runId)
 	}
-	namespace, err := r.GetNamespaceFromRunID(runId)
+	// TODO(gkcalat): consider using run.Namespace after migration logic will be available.
+	namespace, err := r.getNamespaceFromRunId(runId)
 	if err != nil {
-		return util.Wrap(err, "Retry run failed")
+		return util.Wrapf(err, "Failed to retry run %s due to error fetching its namespace", runId)
 	}
 
-	if runDetail.WorkflowSpecManifest != "" && runDetail.WorkflowRuntimeManifest == "" {
-		return util.NewBadRequestError(errors.New("workflow cannot be retried"), "Workflow must be Failed/Error to retry")
+	if run.PipelineSpec.WorkflowSpecManifest != "" && run.RunDetails.WorkflowRuntimeManifest == "" {
+		return util.NewBadRequestError(util.NewInvalidInputError("Workflow manifest cannot be empty"), "Failed to retry run %s due to error fetching workflow manifest", runId)
 	}
-	if runDetail.PipelineSpecManifest != "" {
-		return util.NewBadRequestError(errors.New("workflow cannot be retried"), "Workflow must be with v1 mode to retry")
+	workflowManifest := run.RunDetails.WorkflowRuntimeManifest
+	if workflowManifest == "" {
+		workflowManifest = run.PipelineSpec.WorkflowSpecManifest
 	}
-	execSpec, err := util.NewExecutionSpecJSON(util.ArgoWorkflow, []byte(runDetail.WorkflowRuntimeManifest))
+	execSpec, err := util.NewExecutionSpecJSON(util.ArgoWorkflow, []byte(workflowManifest))
 	if err != nil {
-		return util.NewInternalServerError(err, "Failed to retrieve the runtime pipeline spec from the run")
+		return util.NewInternalServerError(err, "Failed to retry run %s due to error parsing the workflow manifest", runId)
 	}
 
 	if err := execSpec.Decompress(); err != nil {
-		return util.NewInternalServerError(err, "Failed to decompress workflow")
+		return util.NewInternalServerError(err, "Failed to retry run %s due to error decompressing execution spec", runId)
 	}
 
 	if err := execSpec.CanRetry(); err != nil {
-		return err
+		return util.NewInternalServerError(err, "Failed to retry run %s as it does not allow reties", runId)
 	}
 
 	newExecSpec, podsToDelete, err := execSpec.GenerateRetryExecution()
 	if err != nil {
-		return util.Wrap(err, "Retry run failed.")
+		return util.Wrapf(err, "Failed to retry run %s", runId)
 	}
 
 	if err = deletePods(ctx, r.k8sCoreClient, podsToDelete, namespace); err != nil {
-		return util.NewInternalServerError(err, "Retry run failed. Failed to clean up the failed pods from previous run.")
+		return util.NewInternalServerError(err, "Failed to retry run %s due to error cleaning up the failed pods from the previous attempt", runId)
 	}
 
 	// First try to update workflow
-	updateError := r.updateWorkflow(ctx, newExecSpec, namespace)
+	// If fail to get the workflow, return error.
+	latestWorkflow, updateError := r.getWorkflowClient(namespace).Get(ctx, newExecSpec.ExecutionName(), v1.GetOptions{})
```

Comment:
Looking at the diff, the original call was `updateWorkflow`, and the new call is `getWorkflowClient` with return `updateError`,It's not very obvious if this change is correct. @Linchin to take a look as she's working on retry API.

In general, I'd suggest we don't introduce behavior change everywhere but limit behavior changes to pipeline service only.

Reply:
Looks like the logic inside `r.updateWorkflow()` (See [here](https://github.com/kubeflow/pipelines/blob/874e3d3e9136bb16c59cfcf4a3680de40513451c/backend/src/apiserver/resource/resource_manager.go#L668)) is moved to the outside. It seems correct, but I'm not sure if it's necessary.